Add ruby 3.1

jackorp opened the pull request:
Disabled for: RHEL7 (won't ship there, maybe it is better to delete the related files?)
RHEL8, RHEL9 -- not yet available
Fedora: requires s2i-base of at least F36, I could not find such image on quay.io.

Available for: C9S

There is an issue with rails-ex related tests (== from_dockerfile*), other pass: https://gist.github.com/jackorp/0e131fe3b979c48098347ae6380d7df3#file-c9s_ruby31-log-L2036

jackorp commented:
@pvalena JFYI

pvalena commented:
I have tested only the c9s ... with _updated_ rails-ex app as well.

Let's refresh the Gemfile.lock first, though. https://github.com/sclorg/rails-ex/pull/150

(I'm just testing the final version, it seems.)

pvalena commented:
Pre rails-ex refresh -- RHEL-9 seems to be overloaded or broken :).

RHEL-8: still pending (will update)

pvalena commented:
post rails-ex refresh:

[test] [test-openshift]

pkubatrh commented:
RHEL9 fails at subscription step because we are using 9.1 as the host for some reason

@phracek We really should not be using pre-release RHEL versions in production

jackorp commented:
> RHEL-7: ''Your PostgreSQL is too old. Either install an older version of this gem or''

Not really sure how to remedy this situation. The best bet (if we want to keep running that suite) would be IMHO to restrict the `pg` gem version to something that works with RHEL 7's PGSQL, which would be probably the gem version 1.2.3 that was there until now.

phracek commented:
@jackorp @pvalena I guess, ruby-3.1 will be available only for RHEL8, RHEL9, and Fedora the other Dockerfiles like Dockerfile, Dockerfile.rhel7 can disappear for version 3.1

phracek commented:
> RHEL9 fails at subscription step because we are using 9.1 as the host for some reason
> 
> @phracek We really should not be using pre-release RHEL versions in production

This will be fixed by https://github.com/sclorg/s2i-ruby-container/pull/415

phracek commented:
[test-all]

jackorp commented:
> and Fedora the other Dockerfiles like Dockerfile, Dockerfile.rhel7 can disappear for version 3.1

Ok, thanks for the info. I wasn't sure what to do with them. However, Fedora 36/37 and probably even 38 (at least until upstream releases new Ruby in the Fedora 38 timeframe) have Ruby 3.1 at least as an RPM, so please clarify, is this a decision that the Fedora container won't be supported/shipped even though Ruby is available in the repositories?

Follow up on the deletion of Dockerfiles. Should I also remove the C9S Dockerfile, or is that to be kept?

phracek commented:
> Ok, thanks for the info. I wasn't sure what to do with them. However, Fedora 36/37 and probably even 38 (at least until upstream releases new Ruby in the Fedora 38 timeframe) have Ruby 3.1 at least as an RPM, so please clarify, is this a decision that the Fedora container won't be supported/shipped even though Ruby is available in the repositories?

Dockerfile.fedora can be used. Only use `FROM quay.io/fedora/s2i-base:36`. The latest F37 and F38 are not supported yet for s2i-base-container.

phracek commented:
> Follow up on the deletion of Dockerfiles. Should I also remove the C9S Dockerfile, or is that to be kept?

If CentOS Stream 9 supports ruby-31, then we can use it.

pvalena commented:
can you build and push to `quay.io/s2i-ruby/devel:3.1` for testing?

pvalena commented on 3.1/README.md:
```diff
@@ -0,0 +1,229 @@
+Ruby 3.0 container image
```
3.1

pvalena reviewed (changes requested):
There's quite a few mentions of 3.0:

```
$ cd 3.1 ; grep ....
README.md:Ruby 3.0 container image
README.md:This container image includes Ruby 3.0 as a [S2I](https://github.com/openshift/source-to-image) base image for your Ruby 3.0 applications.
README.md:Ruby 3.0 available as container is a base platform for
README.md:building and running various Ruby 3.0 applications and frameworks.
README.md:For this, we will assume that you are using the `ubi8/ruby-30 image`, available via `ruby:3.0` imagestream tag in Openshift.
README.md:Building a simple [ruby-sample-app](https://github.com/sclorg/s2i-ruby-container/tree/master/3.0/test/puma-test-app) application
README.md:    oc new-app ruby:3.0~https://github.com/sclorg/rails-ex.git
root/help.1:.TH Ruby 3.0 container image
root/help.1:This container image includes Ruby 3.0 as a S2I
root/help.1:\[la]https://github.com/openshift/source-to-image\[ra] base image for your Ruby 3.0 applications.
root/help.1:Ruby 3.0 available as container is a base platform for
root/help.1:building and running various Ruby 3.0 applications and frameworks.
root/help.1:For this, we will assume that you are using the \fB\fCubi8/ruby-30 image\fR, available via \fB\fCruby:3.0\fR imagestream tag in Openshift.
root/help.1:\[la]https://github.com/sclorg/s2i-ruby-container/tree/master/3.0/test/puma-test-app\[ra] application
root/help.1:oc new-app ruby:3.0~https://github.com/sclorg/rails-ex.git
s2i/bin/assemble:    # This is new in version 3.0

# Also:
 grep -r '\-30'
README.md:For this, we will assume that you are using the `ubi8/ruby-30 image`, available via `ruby:3.0` imagestream tag in Openshift.
README.md:podman pull ubi8/ruby-30
README.md:An RHEL7 image `ubi8/ruby-30` is used in this example.
README.md:FROM ubi8/ruby-30
README.md:FROM ubi8/ruby-30
root/help.1:For this, we will assume that you are using the \fB\fCubi8/ruby-30 image\fR, available via \fB\fCruby:3.0\fR imagestream tag in Openshift.
root/help.1:podman pull ubi8/ruby-30
root/help.1:An RHEL7 image \fB\fCubi8/ruby-30\fR is used in this example.
root/help.1:FROM ubi8/ruby-30
root/help.1:FROM ubi8/ruby-30
cccp.yml:job-id: ruby-30-centos7

```

pvalena commented:
I've tested this with rails-ex, otherwise LGTM.

jackorp commented:
> There's quite a few mentions of 3.0

Fixed.

Also added 3.1 to the main README.md, however, it often mentions RHEL/CentOS 7, so I left Ruby 3.0 in those parts there for now.

Not sure what to do with `cccp.yml` as it lists centos7.

Otherwise, `make test OS=c9s VERSIONS="3.0 3.1"` passed

phracek commented:
[test-all]

phracek commented:
[test-all]

phracek reviewed (changes requested):
Please add `.exclude-$OS` files to directory 3.1 which will not be tested.

The build is failing on RHEL8:
```
Red Hat Universal Base Image 8 (RPMs) - AppStre  14 MB/s | 3.0 MB     00:00    
Red Hat Universal Base Image 8 (RPMs) - CodeRea  89 kB/s |  20 kB     00:00    
Error: Problems in request:
missing groups or modules: ruby:3.1
Error: error building at STEP "RUN yum -y module enable ruby:$RUBY_VERSION &&     INSTALL_PKGS="     libffi-devel     ruby     ruby-devel     rubygem-rake     rubygem-bundler     ruby-bundled-gems     redhat-rpm-config     " &&     yum install -y --setopt=tsflags=nodocs ${INSTALL_PKGS} &&     yum -y clean all --enablerepo='*' &&     rpm -V ${INSTALL_PKGS}": error while running runtime: exit status 1
make[1]: Leaving directory '/tmp/sclorg/s2i-ruby-container'
```
Please add to 3.1 directory file called `.devel-repo-rhel8`, like is mentioned https://github.com/sclorg/container-common-scripts/blob/master/build.sh#L123 in order to test it against DEVEL_REP for RHEL8.

jackorp commented:
I have pushed the devel repo file for RHEL 8 and RHEL 9 as I expect both RHELs to have Ruby 3.1 in some devel repo. However, I am not sure if the "devel repo" is referencing some internal preview that would contain Ruby 3.1. If that is not the case, then excluding the RHEL Ruby 3.1 tests would be the way to go for now.

pkubatrh commented:
>Not sure what to do with cccp.yml as it lists centos7.

Remove it. It was only used for CCCP to push Centos images into Centos registries. But we are doing that ourselves now, and ruby 3.1 wont be coming to RHSCL anyway.

>I have pushed the devel repo file for RHEL 8 and RHEL 9 as I expect both RHELs to have Ruby 3.1 in some devel repo. 

It should be in the RHEL8 devel repo. I do not think we have this option active for RHEL9 though (at least I could not find a reference to it). @phracek can you check?

[test-all]

jackorp commented:
Looking at some of the RHEL 9 tests, those failed on installing ruby:3.1. Let's exclude that from testing for now.

pkubatrh commented:
[test-all]

phracek commented:
@jackorp Devel-repo for RHEL9 is added to Testing Farm. Can you please add the `.devel-repo-rhel9` file and remove the `.exclude-rhel9` file?

jackorp commented:
@phracek done.

phracek commented:
[test-all]

phracek commented:
[test-all]

phracek commented:
RHEL7 - ruby-2.7 and ruby-3.0 is failed on:

```
[91m[0mLoaded plugins: ovl, product-id, search-disabled-repos, subscription-manager
[91mCould not retrieve mirrorlist https://git.app.eng.bos.redhat.com/git/RH_Software_Collections.git/plain/Containers/osbs-repos-signed-pkgs/pulp_multiarch/x86_64/optional error was
14: curl#60 - "Peer's Certificate issuer is not recognized."
[0m[91m

```

phracek commented:
[test-all]

zmiklank commented:
[test-all]

phracek commented:
@jackorp Can you please rebase common submodule?

jackorp commented:
rebased.

phracek commented:
RHEL7 failures are fixed by #421

phracek commented:
Tests should be passed for now, excluding CentOS 7 which has an Infra problem.

Fingers-crossed.

[test-all]

jackorp commented:
Looks like some changes didn't propagate through git rebase correctly. I'll try to fix it up.

jackorp commented:
Ok, now the changeset should be in sync with the master branch.

phracek commented:
[test-all]

phracek commented:
Openshift 4 tests will be fixed by https://github.com/sclorg/s2i-ruby-container/issues/422.

phracek reviewed (approved):
LGTM. Thanks.
